unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -125,30 +129,37 @@ public UTF8String substring(final int start, final int until) {
   }
 
   public boolean contains(final UTF8String substring) {
+    if (substring == null) return false;
     final byte[] b = substring.getBytes();
     if (b.length == 0) {
       return true;
     }
 
     for (int i = 0; i <= bytes.length - b.length; i++) {
-      // TODO: Avoid copying.
-      if (bytes[i] == b[0] && Arrays.equals(Arrays.copyOfRange(bytes, i, i + b.length), b)) {
+      if (bytes[i] == b[0] && startsWith(b, i)) {
         return true;
       }
     }
     return false;
   }
 
+  private boolean startsWith(final byte[] prefix, int offset) {
```

Comment:
At the first sight, I thought the `offset` is for the start point of the `prefix`. However, if the `offset` represents the start pos of the current `UTF8String.bytes` in the comparison, probably we need to change the method name to `contains` or something, and switch the `offset` and `prefix`.